doctest/doctest.h

```diff
@@ -311,6 +317,12 @@ DOCTEST_MSVC_SUPPRESS_WARNING(26812) // Prefer 'enum class' over 'enum'
 #define DOCTEST_ALIGNMENT(x) __attribute__((aligned(x)))
 #endif
 
+// Work-around for the warning: 'routine is both "inline" and "noinline"'
```

Comment:
I suppose this is because of this line: `inline DOCTEST_NOINLINE void der::f()` (here: https://github.com/doctest/doctest/blob/master/doctest/doctest.h#L1872 )

The reason it's both inline & noinline is because it needs to be a weak symbol (inline) and thus be able to be defined in multiple translation units (if coming from a header file) but we still don't want it to be inline-able (as a compiler optimization) wherever it's called (can't remember why I made it non-inline-able though...).

Other compilers don't have a problem with it - perhaps the ICC warning should be suppressed in another way and not by completely undefining `DOCTEST_NOINLINE`? `DOCTEST_NOINLINE` helps a lot with compile speeds.

Reply:
The error line as follows:
```
/common/doctest.h(1251): error #2196: routine is both "inline" and "noinline"
          DOCTEST_DO_BINARY_EXPRESSION_COMPARISON(==, " == ", DOCTEST_CMP_EQ) //!OCLINT bitwise operator in conditional
```

Reply:
judging by the documentation here:
https://community.intel.com/t5/Intel-C-Compiler/Where-is-the-online-documentation-for-suppressing-warnings/m-p/1111943

seems that those lines with `DOCTEST_DO_BINARY_EXPRESSION_COMPARISON` can be surrounded with:
```c++
#if DOCTEST_ICC
#pragma warning (push)
#pragma warning (disable:2196)
#endif // ICC
...
#if DOCTEST_ICC
#pragma warning (pop)
#endif // ICC
```

But I can't test this... Can you?

Reply:
It's weird, but warning still exists until I remove `#pragma warning (pop)` stuff.

Reply:
My understanding is that Intel C++ Compiler's warning suppression plays weirdly with macros. E.g. it's possible that suppression does not work at the macro expansion point (because the warning is actually inside a macro) and simultaneously does not work at the macro definition point (because it's expanded elsewhere). For example, if I take the latest released doctest 2.4.7 and [the factorial example](https://github.com/doctest/doctest/blob/a384864346f48c7ff5608051c176e7bd505712ae/doc/markdown/tutorial.md#a-simple-example), disable 2196 for the whole TU and enable it at the very bottom, I still get warnings:

```
a.cpp(13): warning #2196: routine is both "inline" and "noinline"

doctest.h(1320): warning #2196: routine is both "inline" and "noinline"
          DOCTEST_DO_BINARY_EXPRESSION_COMPARISON(==, " == ", DOCTEST_CMP_EQ) //!OCLINT bitwise operator in conditional
          ^
```